Manage rule approval requirements for tape sites

germanfgv opened the pull request:
Fixes #10104 

#### Status
ready

#### Description
Adds requireApproval method to Rucio wrapper. This method returns whether a site needs rule approval according to site attributes. RucioInjector uses this new method during the container rule creation process to add the --ask-approval flag to according to site needs.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Builds upon #10006 

#### External dependencies / deployment changes

germanfgv committed:
Add requiresApproval method to Rucio class. Manage RSE rule approval requirements in insertContainerRules

germanfgv committed:
Unit test for requiresApproval

germanfgv commented:
@amaltaro Is there a problem with automatic testing?

amaltaro commented:
Those RSEs don't exist in the Rucio Integration server, yet. You need to append _Test to them...
Review is coming soon.

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +796,26 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+        :param res: Rucio RSE name
+        :returns: True if the RSE requires approval to write (rule property)
+        """
+        
```
German, let me suggest to remove this blank line. According to the PEP convention, there shouldn't be blank lines after the docstring (and I agree that its absence makes the code more readable). If you want, see:
https://www.python.org/dev/peps/pep-0257/

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +796,26 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+        
+        try:
+            attrs = self.cli.list_rse_attributes(rse)
+        except RSENotFound as exc:
```
How about we don't throw this exception, instead just make a log error record, and return a sensible default value as you do in the line below?

amaltaro replied:
And changing my mind here, perhaps we can through the generic WMRucioException with the proper error message, instead of creating a new one.

germanfgv replied:
For this we may use `True` as default. I was not sure of what should be the response to trying to create a rule to an un-existent site. I'll make the changes as you suggest.

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +796,26 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+            msg = "RSE not found in Rucio: {}. Error: {}".format(rse, str(exc))
+            raise WMRucioRSENotFound(msg)
+
+        requiresApproval = attrs.get('requires_approval', False)
```
I'm inclined to think that having a default value to `True` would be safer, no?

germanfgv replied:
The thing is, sites that don't require approval don't have a `requires_approval` attribute. At least I have never seen 'requires_approval=False'. It is either 'requires_approval=True' or not 'requires_approval' at all.

amaltaro replied:
Gotcha! Then default to `False` makes sense.

amaltaro reviewed (changes requested):
If you agree with not throwing that exception, then the new exception can be removed as well.

germanfgv commented on src/python/WMComponent/RucioInjector/RucioInjectorPoller.py:
```diff
@@ -538,6 +538,16 @@ def insertContainerRules(self):
                 rseName = "%s_Test" % rseName
+            
+            #Checking whether we need to ask for rule approval
+            try:
```
@amaltaro according to discussion over `requiresApproval` method, I'm adding this try clause here. I guess I could instead call the method inside the previously existing try clause (ln 554) but I like it better this way. What do you think?

amaltaro replied:
This try/except here looks good. Most of the error message is already encoded in the exception though, so the error message here can likely be shorter and append str(exc) etc.

germanfgv replied:
I shortened the message

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +789,24 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+        _requiresApproval_
+
+        Returns wether or not the specified RSE requires rule approval.
+        :param res: Rucio RSE name
```
can you mention the data type here as well? E.g.: "string with the Rucio RSE name"

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +789,24 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+            attrs = self.cli.list_rse_attributes(rse)
+        except RSENotFound as exc:
+            msg = "Error retrieving attributs for RSE: {}. Error: {}".format(rse, str(exc))
+            raise WMRucioException(msg)
```
I'd replace the else: ... by something like:
```
return attrs.get('requires_approval', False)
```

if we got to this point, then there was no exception.

germanfgv commented:
@amaltaro I think I have addressed your concerns. Also, I tested this with Tier0 agent and seems ti be working as expected. I think this is ready to be merged. Do you (or @todor-ivanov ) have additional comments?

amaltaro commented on src/python/WMCore/Services/Rucio/Rucio.py:
```diff
@@ -789,6 +789,22 @@ def pickRSE(self, rseExpression='rse_type=TAPE\cms_type=test', rseAttribute='ddm
+        try:
+            attrs = self.cli.list_rse_attributes(rse)
+        except RSENotFound as exc:
+            msg = "Error retrieving attributs for RSE: {}. Error: {}".format(rse, str(exc))
```
Can you please fix this typo "attributs"?

germanfgv replied:
Done!

amaltaro replied:
Thanks!

amaltaro reviewed (changes requested):
German, changes look good to me. Can you just fix that typo pointed out in the code?
